File: 2024/May/5-16 2331 Evaluate Boolean Binary Tree/EvaluateTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def evaluateTree(self, root: Optional[TreeNode]) -> bool:
        def evaluate(node: TreeNode) -> bool:
            match node.val:
                case 0:
                    return False
                case 1:
                    return True
                case 2:
                    return evaluate(node.left) or evaluate(node.right)
                case 3:
                    return evaluate(node.left) and evaluate(node.right)
        
        return evaluate(root)
        

File: 2024/May/5-16 2331 Evaluate Boolean Binary Tree/EvaluateTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean evaluateTree(TreeNode root) {
        return evaluate(root);
    }

    private boolean evaluate(TreeNode node) {
        switch(node.val) {
            case 0:
                return false;
            case 1:
                return true;
            case 2:
                return evaluate(node.left) || evaluate(node.right);
            case 3:
                return evaluate(node.left) && evaluate(node.right);
            default:
                return false;
        }
    }
}